[1.4] Big item use rework.

Mirsario opened the pull request:
## UseItemRework
This PR reworks UseItem to return a nullable boolean, and moves its main (there are multiple) call sites to the ApplyItemTime methods, fixing it not being called with a lot of vanilla items. This also lets GlobalItems properly control whether or not ItemTime should be applied.
```cs
/// Return true if the item actually did something, to force itemTime.
/// Return false to keep itemTime at 0.
/// Return null for vanilla behavior.
```

## UseAnimation hook
This PR adds a UseAnimation hook, which runs at the start of item animations.

## Multiplayer synchronization rework?
UseItem hasn't been working correctly *at ALL*, both in 1.3 and 1.4, due to remote players having very poor item use synchronization. In 1.3, UseItem would be called every frame on remote players, due to a lot of the shoot code never being called for them, and, therefore, itemTime not being ever reset from zero.
I've removed a bit of vanilla item use synchronization and made remote players run more code, based on their inputs, which are always synchronized. This also fixes a bunch of vanilla bugs, like sounds not being played correctly. Quite a radical change, but we're at the last major update of the game, and I think this is worth it, for modders' sake.

## Use Multiplier Hooks
![image](https://user-images.githubusercontent.com/6795251/118977035-67988a00-b97e-11eb-91d8-2e60b36ea9b7.png)
The UseTimeMultiplier and MeleeSpeedMultiplier hooks were horribly confusing:
- UseTimeMultiplier wasn't even a multiplier, but a divisor.
- MeleeSpeedMultiplier could easily cause issues, like making the clockwork assault rifle shoot more or less than 3 projectiles it would shoot in vanilla.
- There wasn't a hook for just animation time.
- Why did MeleeSpeedMultiplier have 'Melee' in its name?

I've turned UseTimeMultiplier into an actual multiplier, and replaced MeleeSpeedMultiplier with two new hooks:
- UseAnimationMultiplier - Multiplies itemAnimation, similarly to UseTimeMultiplier, without a care for possible use count implications.
- UseSpeedMultiplier - Multiplies the item's overall usage speed, affecting even reuseDelay (unlike the other hooks.) BUT, it now additionally prevents issues with wrong use counts. This hook is completely safe to use for status effects, like an overcharge that doubles overall weapon speed. And even if you return 9999 here, the clockwork assault rifle will still shoot 3 shots.

## Player itemTime & itemAnimation desychronization fix
Vanilla has had a really horrible bug in it since the very beginning of the game. With it, every time the player would use an item that has autoReuse on it, the itemAnimation value would start lagging behind itemTime, the offset increasing by 1 every time the item is fired/used. In the video below, you can quite easily notice that the firing sound starts to heavily desync after a while.

https://user-images.githubusercontent.com/6795251/124508971-a3c26580-ddd9-11eb-9e81-32498e1a04d3.mp4

This issue comes from very random execution orders that lie inside the Player.ItemCheck method:
```diff
- 1. Reuse delay is applied
- 2. Item animation is applied if button is pressed
- 3. Item animation is reduced
- 4. Hold / Use styles are invoked
- 5. Item time is reduced
- 6. Item logic applies item time
```
It is resolved by changing the execution orders to the following:
```diff
+ 1. Item animation is reduced
+ 2. Item time is reduced
+ 3. Reuse delay is applied
+ 4. Item animation is applied if button is pressed
+ 5. Item logic applies item time
+ 6. Hold / Use styles are invoked
```
(Thanks @ThomasThePencil for help with this, been quite confusing)

## Other fixes
- Fixed weapons not being able to fire when their itemAnimation is at 1.
- The 2 in `Max(2, ...)` calls around total time methods is now 1.
- Total time methods are now in CombinedHooks class.

Mirsario committed:
Attempted to repair UseItem, made it return nullables. Not perfect.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Fixed an accidental removal of GlobalItem.Instance call.

Mirsario committed:
Fixed EM.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

# Conflicts:
#	ExampleMod/Content/Items/Consumables/PlanteraItem.cs
#	patches/tModLoader/Terraria/ModLoader/ItemLoader.cs

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Experimental: UseAnimation hook.

Mirsario committed:
Fixed compilation after ModItem.UseItem bool? -> bool revert.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Fix EM

Mirsario committed:
Changed ModItem.UseItem back to 'bool?' because I'm forgetful and indecisive.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Fixed EM.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Fixed reuseDelays not working.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

# Conflicts:
#	patches/tModLoader/Terraria/ModLoader/ItemLoader.cs

Mirsario committed:
Fixed a bad merge.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

direwolf420 commented on patches/tModLoader/Terraria/ModLoader/GlobalItem.cs:
```diff
@@ -296,11 +296,17 @@ public abstract class GlobalItem : GlobalType<Item>
-		}
+		public virtual bool? UseItem(Item item, Player player) => null;
+
+		/// <summary>
+		/// Allows you to make things happen when an item's use animation starts.
+		/// </summary>
+		public virtual void UseAnimation(Item item, Player player) { }
```
missing parameter documentation

Mirsario replied:
There's kind of no point in documentation like 'item: The item.'.

direwolf420 commented on patches/tModLoader/Terraria/ModLoader/ItemLoader.cs:
```diff
@@ -728,23 +728,40 @@ GlobalItem Instantiate(GlobalItem g)
+		private static HookList HookUseAnimation = AddHook<Action<Item, Player>>(g => g.UseAnimation);
+		
+		public static void UseAnimation(Item item, Player player) {
+			foreach (var g in HookUseAnimation.arr) {
```
artefact from mergedtesting

```suggestion
			foreach (var g in HookUseAnimation.Enumerate(item.globalItems)) {
```

Mirsario committed:
Update patches/tModLoader/Terraria/ModLoader/ItemLoader.cs

Co-authored-by: direwolf420 <amakejk1@web.de>

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

# Conflicts:
#	patches/tModLoader/Terraria/Player.cs.patch

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Experimental: Making item synchronization be just input-based.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

# Conflicts:
#	patches/tModLoader/Terraria/MessageBuffer.cs.patch
#	patches/tModLoader/Terraria/Player.cs.patch

Mirsario committed:
Replaced MeleeTimeMultiplier with UseAnimationMultiplier and UseSpeedMultiplier;
UseTimeMultiplier is now an actual multiplier instead of a divisor;
Fixed use count issues related to MeleeTimeMultiplier (now UseSpeedMultiplier);

Mirsario committed:
Fixed weapons not being able to fire at the last (or only) use tick.

Mirsario committed:
Fixed inconsistent ammo usage on Clockwork Rifle & Eventide.

Mirsario committed:
Added a comment to a questionable patch.

Mirsario committed:
Removed a debug patch.

Mirsario committed:
Fixed EM.

Mirsario committed:
Fixed rotations not being synchronized in multiplayer.

Mirsario commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -3559,7 +3576,7 @@
  					}
 -					else if (itemTime == item.useTime / 2) {
-+					else if (itemTime == PlayerHooks.TotalUseTime(item.useTime, this, item) / 2) {
++					else if (itemTime == CombinedHooks.TotalUseTime(item.useTime, this, item) / 2) {
```
True, but I need to check if that field is updated properly first.

Mirsario committed:
Removed harmful int casts.

Mirsario committed:
Replaced some TotalAnimationTime calls with itemAnimationMax

Mirsario committed:
Used itemTimeMax in a few places.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Used itemTimeMax field in EM.

Mirsario committed:
Fixed a bug with some items not being usable.

Mirsario committed:
Changed how item anim compensations work, fixed clashes with modifiers.

Mirsario committed:
Fixed item animation blinking, added new properties to Player.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

# Conflicts:
#	ExampleMod/Content/Items/Tools/ExampleMagicMirror.cs
#	patches/tModLoader/Terraria/GameInput/PlayerInput.cs.patch
#	patches/tModLoader/Terraria/Player.cs.patch

Mirsario committed:
Merge commit 'a1e6523d2613b3e1480920557bfad5524edebf2c' into 1.4_useitemfix

# Conflicts:
#	patches/tModLoader/Terraria/Item.TML.cs
#	patches/tModLoader/Terraria/Item.cs.patch

Mirsario committed:
"Merged" 1.4 without changes - restores itemTime/itemAnim desync fix that's been moved from the main branch recently.

ThomasThePencil committed:
hypothetical hyper-simplified use time fix

moved the itemTime and itemAnimation decrements to the start of the method

ThomasThePencil committed:
reverted HandleMPItemAnimation change for now and added some explanations

gotos are confusing okay

Mirsario committed:
Merge branch '1.4_useitemfix' into 1.4_simplifiedusetimefix

# Conflicts:
#	patches/tModLoader/Terraria/Player.cs.patch

Mirsario committed:
Forgot to rediff.

Mirsario committed:
Fixed a bad rebase and various issues.

Mirsario committed:
Removed my now-unnecessary and questionable patches.

Mirsario committed:
Made item holding logic execute after item logic.

Mirsario committed:
It's dangerous to go alone. Included a map of itemcheck execution order.

Mirsario committed:
Merge pull request #1601 from ThomasThePencil/1.4_simplifiedusetimefix

[1.4] Hypothetical hyper-simplified use time fix

Mirsario committed:
Reorganized comments, + linked this PR in a confusing portion.

Mirsario committed:
Replaced one more TotalUseTime call with itemTImeMax

Mirsario committed:
Fixed item anim compensations going in the wrong way.

Mirsario committed:
Removed the now-useless InItemAnimation property.

Mirsario committed:
Fixed a rollback in protection modifiers of some Player methods.

Mirsario committed:
Removed ImteTimePrev and ItemAnimationPrev properties.

Mirsario committed:
Fixed another rollback in Player.cs.

Mirsario committed:
Commented a patch.

Mirsario committed:
Fixed ammo use on flamethrowers being different from vanilla.

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

Mirsario committed:
Merge branch '1.4' into 1.4_useitemfix

DarioDaF commented:
Tested `Player.ItemCheck_Inner` "spaghetti" reordering, this action will result in hitboxes being updated after item use because `Hold / Use styles are invoked` gets last in the chain resulting in bug: https://github.com/tModLoader/tModLoader/issues/1721 .
Furthermore removing the reordering (commenting the gotos) will result in collaterally fixing also https://github.com/tModLoader/tModLoader/issues/1770 and https://github.com/tModLoader/tModLoader/issues/1767 related to how feral claw updates stuff.
Empirically in `HandleItemHolding` the `ItemCheck_ApplyHoldStyle` is **never called** for melee weapons (not checked wit other weapons, but seems reasonable to think it's never called ever) because due to reordering `itemAnimation > 0` is always true maybe?
